### PR TITLE
[labs/context] Make `@provide` and `@consume` support optional and private properties.

### LINKED ISSUE #3732: [labs/context] `@provide` forces a non-optional property
### Which package(s) are affected?

Context (@lit-labs/context)

### Description

The code is this:

```ts
  @provide({ context: studentContext })
  @property({ type: Object })
  student?: Student;
```

The typing error I get with this:

```
Argument of type 'Root' is not assignable to parameter of type 'ReactiveElement & Record<"student", Student>'.
  Type 'Root' is not assignable to type 'Record<"student", Student>'.
    Types of property 'student' are incompatible.
      Type 'Student | undefined' is not assignable to type 'Student'.
        Type 'undefined' is not assignable to type 'Student'.ts(2345)
```

If I make the property *not* optional (so remove the question mark), then I get this typing error:

```
Property 'student' has no initializer and is not definitely assigned in the constructor.ts(2564)
```

"Damned if I do, damned if I don't"....

### Reproduction

-

### Workaround

I've tried this, to no avail: `export const studentContext = createContext<Student | undefined>('student');`

If I do the above, then I do get a different typing error:

```
Argument of type 'Root' is not assignable to parameter of type 'ReactiveElement & Record<"student", Student | undefined>'.
  Type 'Root' is not assignable to type 'Record<"student", Student | undefined>'.
    Property 'student' is optional in type 'Root' but required in type 'Record<"student", Student | undefined>'.ts(2345)
```

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

"@lit-labs/context": "^0.2.0",

### Browser/OS/Node environment

`"typescript": "^4.9.5"` with `"strict": true,` in `tsconfig.json`

### PR DESCRIPTION
**Pull request by jpzwarte**
Fixes #3732 

As far as I can tell, the change just works. All tests still pass.

### CODE REVIEW

**Comment by justinfagnani** [comment]
I think the reason I didn't make this change when I did the change for `@consume()` is that there's the question of whether this provides `undefined` for a context key that can't be undefined.

```ts
const fooContext = createContext<string>('foo');

class Parent extends LitElement {
  @provide({context: fooContext})
  foo?: Foo;
  
  render() { return html`<x-child></x-child>`; }
}

class Child extends LitElement {
  @consume({context: fooContext})
  foo?: Foo;
}
```

In this case because we have a context of a `string` and not a `string | undefined`, if would be an error for this `@provide/foo` to provide the value.

But that's difficult to differentiate from this case:

```ts
const fooContext = createContext<string | undefined>('foo');

class Parent extends LitElement {
  @provide({context: fooContext})
  foo?: Foo;
```

where it _would_ be ok to provide a value of `undefined`.

So I think that either:
 1. we can't tell whether to provide a value in this case and we need to keep the current requirement that the field is defined and matches the context type (which you need to include `undefined` if you want that) and we always provide a value, or
 2. we add a test to this suite that ensures that the consumer gets a `number` and not `undefined`, even if we don't set a value like in the other tests. I think we want that test anyway to nail down the intended behavior here.

**Comment by rictic** [comment]
I think we can use a type like this to distinguish between the case where the context allows undefined (in which case an optional field is fine) from the case where it isn't.

```typescript
type LooseRecord<K, V> = 
    V extends undefined ? 
        Partial<Record<K, V>> : 
        Record<K, V>;
```

I'll try it out and add some tests

**Comment by rictic** [comment]
This ended up turning into a slightly bigger change, but I added support for providing and consuming optional properties (with correct type checking), and private/protected properties (with no type checking), and with no changes to the authoring experience. 

Also added type checking tests using `// @ts-expect-error`, as well as tests of our behavior when providing and consuming optional properties.

**Comment by rictic** [comment]
Fixes https://github.com/lit/lit/issues/3883

**Comment by jpzwarte** [comment]
@rictic Thanks for this!

File: packages/labs/context/src/lib/decorators/provide.ts [hunk 1]
```diff
@@ -76,3 +71,46 @@ export function provide<ValueType>({
...
+      ? ReturnValue
+      : {
+          message: 'providing field not assignable to context';
+          context: ContextType;
+          provided: ProvidingType;
+        }
+    : // Next we check whether the object has the property as an optional field
+    Obj extends Partial<Record<Key, infer Providing>>
+    ? // Check assignability again. Note that we have to include undefined
+      // here on the consuming type because it's optional.
```
Comment (AndrewJakubowicz) [inline packages/labs/context/src/lib/decorators/provide.ts:104]: Nit: Should this comment reference the `providing` type?

File: packages/labs/context/src/lib/decorators/provide.ts [hunk 2]
```diff
@@ -76,3 +71,46 @@ export function provide<ValueType>({
...
+      [Providing | undefined] extends [ContextType]
+      ? ReturnValue
+      : {
+          message: 'providing field not assignable to context';
+          context: ContextType;
+          consuming: Providing | undefined;
+        }
+    : // Ok, the field isn't present, so either someone's using consume
```
Comment (AndrewJakubowicz) [inline packages/labs/context/src/lib/decorators/provide.ts:112]: Nit: "using ~consume~ **provide**"

**Review by AndrewJakubowicz — APPROVED** [review]
Wow! There are some advanced types on display here.

TIL: about how to get around [distributive conditional type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types) behavior using square brackets! 

Great use of type only tests which are necessary with such a complex change. The two types:  `FieldMustMatchProvidedType` and `FieldMustMatchContextType` are very similar. Is it possible (or worthwhile) refactoring them into a single type?

File: .changeset/eighty-bananas-drive.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+'@lit-labs/context': patch
+---
+
+Improve type checking of @provide and @consume. We now support optional properties, and public properties should be fully type checked.
+
+This also adds support for using @provide and @consume with protected and private fields, but be aware that do to limitations in TypeScript's experimental decorators, the connection between the context and the property can not be type checked. We'll be able to fix this when standard decorators are released (current ETA TypeScript 5.2 in August 2023).
```
[the remark this answers is not in the document]
Comment (rictic) [reply .changeset/eighty-bananas-drive.md]: * Done
* Done
* Added!

And sent over as https://github.com/lit/lit.dev/pull/1134

File: packages/labs/context/README.md [hunk 4]
```diff
@@ -157,6 +157,12 @@ This solution has a small overhead, in that if a provider is not within the DOM
 
 Note that ContextRoot uses [WeakRefs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef) which are not supported in IE11.
 
+### Protected / Private Properties
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/context/README.md:160]: Done

File: packages/labs/context/src/lib/decorators/consume.ts [hunk 5]
```diff
@@ -69,3 +63,46 @@ export function consume<ValueType>({
...
+    // Note TypeScript requires the return type to be `void|any`
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ): FieldMustMatchProvidedType<Proto, K, ValueType, void | any>;
+};
+
+type FieldMustMatchProvidedType<
+  Obj,
+  Key extends PropertyKey,
+  ProvidedType,
+  ReturnValue
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/context/src/lib/decorators/consume.ts]: Good point, removed

File: packages/labs/context/src/lib/decorators/provide.ts (hunk 1 above)
Comment (rictic) [reply packages/labs/context/src/lib/decorators/provide.ts]: Good catch!

File: packages/labs/context/src/lib/decorators/provide.ts (hunk 2 above)
Comment (rictic) [reply packages/labs/context/src/lib/decorators/provide.ts]: Ah, right! Done.

File: packages/labs/context/src/test/context-provider_test.ts [hunk 6]
```diff
@@ -84,6 +97,16 @@ suite('@consume', () => {
     await consumer.updateComplete;
     assert.strictEqual(consumer.value, 500);
   });
+
+  test('consuming and providing with optional fields', async () => {
+    assert.strictEqual(consumer.optionalValue, undefined);
+    assert.strictEqual(consumer.consumeOptionalWithDefault, undefined);
+    provider.optionalValue = 500;
+    await provider.updateComplete;
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/context/src/test/context-provider_test.ts]: Actually both awaits can be removed. Done

File: packages/labs/context/src/test/type_check_test.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+  if (false as boolean) {
+    console.log(val);
+  }
+}
+
+suite('compilation tests', () => {
+  // this code doesn't need to run, it's tested by tsc
+  if (true as boolean) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/context/src/test/type_check_test.ts:21]: :ratjam:

File: packages/labs/context/src/test/type_check_test.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+);
+
+// to silence linters
+function markAsUsed(val: unknown) {
+  if (false as boolean) {
+    console.log(val);
+  }
+}
+
+suite('compilation tests', () => {
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/context/src/test/type_check_test.ts:17]: Done!

**Pull request merged**